fabric-model-loading-api-v1/src/client/java/net/fabricmc/fabric/mixin/client/model/loading/BakedModelManagerMixin.java

```diff
@@ -75,22 +92,45 @@ private CompletableFuture<BlockStatesLoader.LoadedModels> hookBlockStateModels(C
 		return modelsFuture.thenCombine(eventDispatcherFuture, (models, eventDispatcher) -> eventDispatcher.modifyBlockModelsOnLoad(models));
 	}
 
+	@ModifyArg(method = "reload", at = @At(value = "INVOKE", target = "java/util/concurrent/CompletableFuture.thenApplyAsync(Ljava/util/function/Function;Ljava/util/concurrent/Executor;)Ljava/util/concurrent/CompletableFuture;", ordinal = 1), index = 0)
+	private Function<Void, ?> hookTextureCollect(Function<Void, CompletableFuture<?>> function) {
+		return withModelDispatcher(function);
+	}
+
 	@ModifyArg(method = "reload", at = @At(value = "INVOKE", target = "java/util/concurrent/CompletableFuture.thenComposeAsync(Ljava/util/function/Function;Ljava/util/concurrent/Executor;)Ljava/util/concurrent/CompletableFuture;", ordinal = 0), index = 0)
 	private Function<Void, CompletableFuture<?>> hookModelBaking(Function<Void, CompletableFuture<?>> function) {
-		return v -> {
-			CompletableFuture<ModelLoadingEventDispatcher> future = eventDispatcherFuture;
-
-			if (future == null) {
-				return function.apply(v);
-			}
+		return withModelDispatcher(function);
+	}
 
+	@Unique
+	private <T, R> Function<T, R> withModelDispatcher(Function<T, R> function) {
+		CompletableFuture<ModelLoadingEventDispatcher> future = eventDispatcherFuture;
+		if (future == null) return function;
+		return x -> {
 			ModelLoadingEventDispatcher.CURRENT.set(future.join());
-			CompletableFuture<?> bakingResultFuture = function.apply(v);
-			ModelLoadingEventDispatcher.CURRENT.remove();
-			return bakingResultFuture;
+
+			try {
+				return function.apply(x);
+			} finally {
+				ModelLoadingEventDispatcher.CURRENT.remove();
+			}
 		};
 	}
 
+	@Inject(method = "collect", at = @At(value = "INVOKE", target = "net/minecraft/client/render/model/ReferencedModelsCollector.collectModels()Ljava/util/Map;"))
+	private static void resolveExtraModels(
+			Map<Identifier, UnbakedModel> modelMap, BlockStatesLoader.LoadedModels stateDefinition, ItemAssetsLoader.Result result, CallbackInfoReturnable<?> cir,
+			@Local ReferencedModelsCollector collector
+	) {
+		// We know eventDispatcherFuture is available, as it is required by the item and block models (hookModels).
+		ModelLoadingEventDispatcher.CURRENT.get().getExtraModels().values().forEach(collector::resolve);
```

Comment:
The value of a thread local should never be assumed to be not null. It may not be 100% necessary, but the rest of the implementation already performs such checks, so best to also do that here.